### PR TITLE
DOCS-741: Improving discussion of object deletion

### PR DESCRIPTION
**Pull request by ravindk89**
Staged: http://192.241.195.202:9000/staging/DOCS-741/linux/administration/object-management.html

Partially addresses #741 

Did a bit of cleanup and re-org of the page.

Added some details on delete operations for each subsection.

### CODE REVIEW

File: source/administration/object-management.rst [hunk 1]
```diff
@@ -45,10 +40,25 @@ following:
          2020-01-02-MinIO-Advanced-Deployment-comments.json
          2020-01-04-MinIO-Interview.md
 
-MinIO supports multiple levels of nested directories and objects using a :term:`prefix` structure to support even the most dynamic object storage workloads.
+With the example structure, an administrator would create the ``/images``, ``/videos`` and ``articles`` buckets.
+Client applications would write objects to those buckets using the full "path" to that object, including all intermediate :term:`prefixes <prefix>`.
```
Comment (djwfyi) [inline source/administration/object-management.rst:44]: Consider dropping "would". It's not really necessary.

File: source/administration/object-management.rst [hunk 2]
```diff
@@ -45,10 +40,25 @@ following:
...
 
-MinIO itself does not limit the number of objects that any specific prefix can contain.
-However, hardware and network conditions may show performance impacts with large prefixes.
+MinIO supports multiple levels of nested directories and objects using prefixes to support even the most dynamic object storage workloads.
+Neither clients nor administrators would manually create the intermediate prefixes, as MinIO automatically infers them from the object name.
```
Comment (djwfyi) [inline source/administration/object-management.rst:47]: > There is no need for clients or administrators to manually create prefixes, as MinIO automatically infers them from the object name.

Neither...nor constructions can be hard to follow, at least I find them difficult.

File: source/administration/object-management.rst [hunk 3]
```diff
@@ -59,16 +69,25 @@ However, hardware and network conditions may show performance impacts with large
...
 .. image:: /images/retention/minio-versioning-multiple-versions.svg
    :alt: Object with Multiple Versions
    :align: center
 
-MinIO versioning protects from unintended overwrites and deletions while providing support for "undoing" a write operation.
+For buckets with versioning disabled or suspended, creating an object at a given namespace (``BUCKET/PREFIX/OBJECT.EXT``) overwrites the object permanently.
+Delete operations remove the object with no possibility of recovery ("hard delete").
+
+For buckets with versioning enabled, write operations on an existing object create a new version of that object using the write payload. 
+Delete operations produce a 0-byte "Delete Marker" as the latest version of the object ("soft delete").
```
Comment (djwfyi) [inline source/administration/object-management.rst:84]: > Delete operations produce a 0-byte "Delete Marker" _for_ the latest version of the object.

I think "for" is better here than "as".

---

I feel like this paragraph and the one above might be easier to consume as a table, or maybe select h3s.
But maybe that's a different pass.

File: source/administration/object-management.rst [hunk 4]
```diff
@@ -83,23 +102,31 @@ MinIO supports both :ref:`duration based object retention <minio-object-locking-
    :align: center
    :width: 600px
 
-See :ref:`minio-object-locking` for more complete documentation.
+Delete operations against a WORM-locked object depend on the specific operation:
+
+- Delete operations which do not specify a version ID result in the creation of a "Delete Marker"
+- Delete operations which specify the version ID of a locked object result in an locking error
```
Comment (djwfyi) [inline source/administration/object-management.rst:108]: > ... a locking error

(Indefinite article is wrong here.)

File: source/administration/object-management.rst [hunk 5]
```diff
@@ -83,23 +102,31 @@ MinIO supports both :ref:`duration based object retention <minio-object-locking-
...
+
+You *must* enable object locking during bucket creation.
+Enabling bucket creation also enables :ref:`versioning <minio-bucket-versioning>`.
```
Comment (djwfyi) [inline source/administration/object-management.rst:111]: What? This doesn't make sense.
Bucket creation does not automatically enable versioning.

**Review by djwfyi — CHANGES_REQUESTED** [review]
A few changes, mostly minor nits.
I'd like to take another look after that.

File: source/administration/object-management.rst [hunk 6]
```diff
@@ -83,23 +102,31 @@ MinIO supports both :ref:`duration based object retention <minio-object-locking-
    :align: center
    :width: 600px
 
-See :ref:`minio-object-locking` for more complete documentation.
+Delete operations against a WORM-locked object depend on the specific operation:
+
+- Delete operations which do not specify a version ID result in the creation of a "Delete Marker"
+- Delete operations which specify the version ID of a locked object result in an locking error
+
+You *must* enable object locking during bucket creation.
```
Comment (feorlen) [inline source/administration/object-management.rst:110]: Maybe this is covered in another part of the document, but just seeing this line by itself suggests object locking is required everywhere whether you want it or not. Perhaps somethng like "it may not be added to an existing bucket" at the end.

File: source/administration/object-management.rst [hunk 7]
```diff
@@ -83,23 +102,31 @@ MinIO supports both :ref:`duration based object retention <minio-object-locking-
...
-MinIO Object Lifecycle Management allows creating rules for time or date
-based automatic transition or expiry of objects. For object transition,
-MinIO automatically moves the object to a configured remote storage
-tier. For object expiry, MinIO automatically deletes the object.
+MinIO Object Lifecycle Management allows creating rules for time or date based automatic transition or expiry of objects. 
+For object transition, MinIO automatically moves the object to a configured remote storage tier. 
+For object expiry, MinIO automatically deletes the object.
+
+MinIO applies lifecycle management rules on :ref:`versioned and unversioned buckets <minio-bucket-versioning>` using the same behavior as normal client operations.
+You can specify transition or lifecycle rules that handle either or both the latest object versions and prior object versions.
```
Comment (feorlen) [inline source/administration/object-management.rst:125]: "or both" here feels oddly placed. Maybe "rules that handle the latest object versions, prior object versions, or both" ?

File: source/administration/object-management.rst [hunk 8]
```diff
@@ -83,23 +102,31 @@ MinIO supports both :ref:`duration based object retention <minio-object-locking-
...
 
-MinIO lifecycle management is built for behavior and syntax compatibility with 
-:s3-docs:`AWS S3 Lifecycle Management <object-lifecycle-mgmt.html>`. For
-example, you can export S3 lifecycle management rules and import them into
-MinIO or vice-versa. MinIO uses JSON to describe lifecycle management rules,
-and conversion to or from XML may be required. 
+MinIO lifecycle management is built for behavior and syntax compatibility with :s3-docs:`AWS S3 Lifecycle Management <object-lifecycle-mgmt.html>`. 
+For example, you can export S3 lifecycle management rules and import them into MinIO or vice-versa. 
+MinIO uses JSON to describe lifecycle management rules, and conversion to or from XML may be required.
```
Comment (feorlen) [inline source/administration/object-management.rst:129]: The "and" here seems superfluous to me

File: source/administration/object-management.rst [hunk 9]
```diff
@@ -10,21 +10,16 @@ Object Management
 
 .. _objects:
 
-An :ref:`object <objects>` is binary data, sometimes referred to as a Binary
-Large OBject (BLOB). Blobs can be images, audio files, spreadsheets, or even
-binary executable code. Object Storage platforms like MinIO provide dedicated
-tools and capabilities for storing, retrieving, and searching for blobs. 
+An :ref:`object <objects>` is binary data, sometimes referred to as a Binary Large OBject (BLOB).
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/administration/object-management.rst]: Did some tweaking here.

File: source/administration/object-management.rst (hunk 2 above)
Comment (ravindk89) [reply source/administration/object-management.rst:50]: Rewritten.

File: source/administration/object-management.rst (hunk 3 above)
Comment (ravindk89) [reply source/administration/object-management.rst]: Fixed, and took a pass at a table.

File: source/administration/object-management.rst (hunk 6 above)
Comment (ravindk89) [reply source/administration/object-management.rst]: Tweaked, it is in fact just at bucket creation that you even have the option.

File: source/administration/object-management.rst (hunk 5 above)
Comment (ravindk89) [reply source/administration/object-management.rst]: yes, i had a brain

File: source/administration/object-management.rst (hunk 7 above)
Comment (ravindk89) [reply source/administration/object-management.rst]: I'll take that suggestion

File: source/administration/object-management.rst (hunk 8 above)
Comment (ravindk89) [reply source/administration/object-management.rst]: Tweaked

**Review by ravindk89 — COMMENTED** [review]
@djwfyi @feorlen back to you!

File: source/administration/object-management.rst [hunk 10]
```diff
@@ -10,21 +10,16 @@ Object Management
...
+The term "Binary Large Object" or "blob" is sometimes associated to object storage, although blobs can be anywhere from a few bytes to several terabytes in size.
+Object Storage platforms like MinIO provide dedicated tools and capabilities for storing, listing, and retrieving objects using a standard S3-compatible API. 
 
 .. _buckets:
 
 MinIO Object Storage uses :ref:`buckets <buckets>` to organize objects. 
-A bucket is similar to a folder or directory in a filesystem, where each
-bucket can hold an arbitrary number of objects. MinIO buckets provide the 
-same functionality as AWS S3 buckets.
+A bucket is similar to a folder or directory in a filesystem, where each bucket can hold an arbitrary number of objects.
```
Comment (djwfyi) [inline source/administration/object-management.rst:20]: would it be worth adding something like 
> A bucket is similar to a **top-level** folder or directory in a filesystem...

to distinguish from prefixes?

File: source/administration/object-management.rst [hunk 11]
```diff
@@ -59,16 +71,48 @@ However, hardware and network conditions may show performance impacts with large
...
+     - Versioning Disabled | Suspended
+
+   * - ``PUT`` (Write)
+     - Create a new full version of the object as the "latest" and assign a unique version ID
+     - Create the object with overwrite on namespace match.
+
+   * - ``GET`` (Read)
+     - Retrieve the latest version of the object by default
+
+       Supports retrieving a specific object by version ID.
```
Comment (djwfyi) [inline source/administration/object-management.rst:101]: Maybe 
> Supports retrieving any object version by version ID

**Review by djwfyi — APPROVED** [review]
Minor suggestions to consider.
The table is 👍 👍 

LGTM otherwise.

**Review by feorlen — APPROVED** [review]
LGTM

File: source/administration/object-management.rst (hunk 11 above)
Comment (ravindk89) [reply source/administration/object-management.rst:101]: :thinking: i'm trying to avoid the double 'version" here. But I see the concern, so meh.
